### ISSUE TITLE
Compile-AppInNAVContainer: Download symbols for non-Microsoft dependent app?

### ISSUE THREAD

**Issue by buzzwick** [microsoft/navcontainerhelper#2333]
All of our deployments are on-premise, so as far as I understand the HelloWorld CI/CD scripts are not an option for automating builds. I have put together a script that at at least gets as far as building a container and pulling down the app code, using a self-hosted agent. Compile-AppInNAVContainer downloads symbols from the MS Application, System, and Base. But when it finds a dependency in our app to our own base app, it continues to look for those symbols at Microsoft, and fails. 

I have tried multiple strategies, but since all of the folders used for symbols are temporary and do not exist until the script is run, or are not available to the container, I have had no success. 

Question: is it possible to either redirect Compile-App for downloading symbols, or to make our base app symbols available to the script?

Script:
# Initial pipeline to build & produce .app file
# add a comment line
trigger:
- master

pool:
  name: Default

variables:
  image_name: mcr.microsoft.com/businesscentral
  container_name: Build
  company_name: xxxxxxxx
  user_name: admin
  password: xxxxxxxx
  license_file: '"C:\_xxx Licenses\v19.flf"'

stages:
- stage: build
  displayName: Build
  jobs:
  - job: Build
    pool:
      name: Default
    steps:
      - task: PowerShell@1
        displayName: Install BCContainerHelper module
        inputs:
          scriptType: inlineScript
          inlineScript: >
            Install-Module -Name BCContainerHelper  -force;

      - task: PowerShell@1    
        displayName: Create build container
        inputs:
          scriptType: inlineScript
          inlineScript: > 
            Import-Module Bccontainerhelper;
            $artifactUrl = Get-BcArtifactUrl -type 'OnPrem' -country 'us' -version '19.0.29894.30693';
            $Credential = [PSCredential]::new('$(user_name)',(ConvertTo-SecureString '$(password)' -AsPlainText -Force));
            New-BcContainer -accept_eula -accept_outdated -containerName '$(container_name)' -auth UserPassword -credential $Credential -image $(image_name) -process hyperv -artifacturl $artifactUrl -licenseFile $(license_file) -doNotExportObjectsToText -includeTestToolkit -includeTestLibrariesOnly -restart no -shortcuts None -useBestContainerOS   -updateHosts
      - task: PowerShell@1
        displayName: Copy source into container folder
        inputs:
          scriptType: inlineScript
          inlineScript: >
            $SourceDir = 'C:\ProgramData\BcContainerHelper\Extensions\$(container_name)\Source';
            New-Item $SourceDir -ItemType Directory;
            Copy-Item '$(Build.SourcesDirectory)\*' $SourceDir -Recurse -Force;
      - task: PowerShell@1
        displayName: Compile app
        inputs:
          scriptType: inlineScript
          inlineScript: >
            Import-Module BcContainerHelper;
            $SourceDir = 'C:\ProgramData\BcContainerHelper\Extensions\$(container_name)\Source';
            $Credential = [PSCredential]::new('$(user_name)',(ConvertTo-SecureString '$(password)' -AsPlainText -Force));
            Compile-AppInNavContainer -containerName '$(container_name)' -appProjectFolder $SourceDir -credential $Credential -AzureDevOps -FailOn 'error';

Output--Container Build, Copy Source: 

`2022-02-27T22:54:04.7443420Z ##[section]Starting: Create build container
2022-02-27T22:54:04.7750569Z ==============================================================================
2022-02-27T22:54:04.7751027Z Task         : PowerShell
2022-02-27T22:54:04.7751434Z Description  : Run a PowerShell script
2022-02-27T22:54:04.7751756Z Version      : 1.2.3
2022-02-27T22:54:04.7752054Z Author       : Microsoft Corporation
2022-02-27T22:54:04.7752488Z Help         : [More Information](https://go.microsoft.com/fwlink/?LinkID=613736)
2022-02-27T22:54:04.7752975Z ==============================================================================
2022-02-27T22:54:04.7818425Z ##[command]. 'C:\Users\xxxxxx\AppData\Local\Temp\fbfbcead-2c03-4774-a841-ddf3c0529bcb.ps1' 
2022-02-27T22:54:05.6638178Z BcContainerHelper version 3.0.2
2022-02-27T22:54:05.6738851Z BcContainerHelper emits usage statistics telemetry to Microsoft
2022-02-27T22:54:08.6364505Z WARNING: The names of some imported commands from the module 'Bccontainerhelper' include unapproved verbs that might 
2022-02-27T22:54:08.6365615Z make them less discoverable. To find the commands with unapproved verbs, run the Import-Module command again with the 
2022-02-27T22:54:08.6366288Z Verbose parameter. For a list of approved verbs, type Get-Verb.
2022-02-27T22:54:12.1911719Z BcContainerHelper is version 3.0.2
2022-02-27T22:54:12.1919695Z BcContainerHelper is running as administrator
2022-02-27T22:54:12.1944544Z Host is Microsoft Windows 10 Pro - 21H1
2022-02-27T22:54:12.7116680Z Docker Client Version is 20.10.12
2022-02-27T22:54:12.7146833Z Docker Server Version is 20.10.12
2022-02-27T22:54:15.7875288Z Fetching all docker images
2022-02-27T22:54:16.5712509Z Fetching all docker volumes
2022-02-27T22:54:17.2814025Z ArtifactUrl and ImageName specified
2022-02-27T22:54:17.9060277Z INFO: Windows 10 21H1/21H2 images are not yet available, using 2004 as these are found to work better than 20H2 on 21H1/21H2
2022-02-27T22:54:17.9221262Z WARNING: Unable to find matching generic image for your host OS. Using mcr.microsoft.com/businesscentral:10.0.19041.1415
2022-02-27T22:54:18.9273151Z Image mcr.microsoft.com/businesscentral:onprem-19.0.29894.30693-us already exists
2022-02-27T22:54:20.6333777Z Using image mcr.microsoft.com/businesscentral:onprem-19.0.29894.30693-us
2022-02-27T22:54:21.3193016Z Creating Container Build
2022-02-27T22:54:21.3291297Z Style: onprem
2022-02-27T22:54:21.3298698Z Multitenant: No
2022-02-27T22:54:21.3322630Z Version: 19.0.29894.30693
2022-02-27T22:54:21.3349436Z Platform: 19.0.29884.30666
2022-02-27T22:54:21.3363123Z Generic Tag: 1.0.2.1
2022-02-27T22:54:21.3400049Z Container OS Version: 10.0.19041.1415 (2004)
2022-02-27T22:54:21.3409129Z Host OS Version: 10.0.19043.1526 (21H1)
2022-02-27T22:54:21.3482968Z WARNING: Host OS is 21H1 and Container OS is 2004, defaulting to process isolation. If you experience problems, add -isolation hyperv.
2022-02-27T22:54:21.3490191Z Using process isolation
2022-02-27T22:54:21.3539524Z Using locale en-US
2022-02-27T22:54:21.3598177Z Disabling the standard eventlog dump to container log every 2 seconds (use -dumpEventLog to enable)
2022-02-27T22:54:21.3672572Z Using license file C:\_Licenses\v19.flf
2022-02-27T22:54:21.4020351Z Files in C:\ProgramData\BcContainerHelper\Extensions\Build\my:
2022-02-27T22:54:21.4106833Z - AdditionalOutput.ps1
2022-02-27T22:54:21.4116139Z - license.flf
2022-02-27T22:54:21.4126447Z - MainLoop.ps1
2022-02-27T22:54:21.4135740Z - SetupVariables.ps1
2022-02-27T22:54:21.4147404Z - updatehosts.ps1
2022-02-27T22:54:21.4156770Z Creating container Build from image mcr.microsoft.com/businesscentral:onprem-19.0.29894.30693-us
2022-02-27T22:54:22.1663666Z f5fa32ea5dcd0a59d1d56fcc41afd49ca74900cf663502942ecee8399b7adf0c
2022-02-27T22:54:26.5653119Z Waiting for container Build to be ready
2022-02-27T22:54:28.3788833Z Initializing...
2022-02-27T22:54:28.3789366Z Setting host.docker.internal to 192.168.1.184 in container hosts file (copy from host hosts file)
2022-02-27T22:54:28.3790051Z Setting gateway.docker.internal to 192.168.1.184 in container hosts file (copy from host hosts file)
2022-02-27T22:54:28.3792041Z Setting kubernetes.docker.internal to 127.0.0.1 in container hosts file (copy from host hosts file)
2022-02-27T22:54:28.3792608Z Setting host.containerhelper.internal to 172.31.48.1 in container hosts file
2022-02-27T22:54:28.3793005Z Starting Container
2022-02-27T22:54:28.3793318Z Hostname is Build
2022-02-27T22:54:30.4561854Z PublicDnsName is Build
2022-02-27T22:54:30.4562154Z Using NavUserPassword Authentication
2022-02-27T22:54:32.6616223Z Starting Local SQL Server
2022-02-27T22:54:32.6616564Z Starting Internet Information Server
2022-02-27T22:54:34.4532721Z Creating Self Signed Certificate
2022-02-27T22:54:34.4533280Z Self Signed Certificate Thumbprint 0AD2B9C92DBFEF2EC222FCA32F25B671CDA9FC3C
2022-02-27T22:54:34.4533843Z DNS identity Build
2022-02-27T22:54:34.4534298Z Modifying Service Tier Config File with Instance Specific Settings
2022-02-27T22:54:59.7743275Z Starting Service Tier
2022-02-27T22:54:59.7743616Z Registering event sources
2022-02-27T22:55:05.3015110Z Creating DotNetCore Web Server Instance
2022-02-27T22:55:05.3015916Z Using license file 'c:\run\my\license.flf'
2022-02-27T22:55:05.3016535Z Import License
2022-02-27T22:55:07.0261345Z Creating http download site
2022-02-27T22:55:07.0261755Z Setting SA Password and enabling SA
2022-02-27T22:55:07.0262110Z Creating admin as SQL User and add to sysadmin
2022-02-27T22:55:10.4350905Z Creating SUPER user
2022-02-27T22:55:10.4351396Z Container IP Address: 172.31.59.129
2022-02-27T22:55:10.4351692Z Container Hostname  : Build
2022-02-27T22:55:10.4351962Z Container Dns Name  : Build
2022-02-27T22:55:10.4352380Z Web Client          : http://Build/BC/
2022-02-27T22:55:10.4353111Z Dev. Server         : http://Build
2022-02-27T22:55:10.4353602Z Dev. ServerInstance : BC
2022-02-27T22:55:10.4354016Z Setting Build to 172.31.59.129 in host hosts file
2022-02-27T22:55:10.4354250Z 
2022-02-27T22:55:10.4354524Z Files:
2022-02-27T22:55:10.4354967Z http://Build:8080/ALLanguage.vsix
2022-02-27T22:55:10.4355210Z 
2022-02-27T22:55:10.4355520Z Container Total Physical Memory is 31.9Gb
2022-02-27T22:55:10.4355872Z Container Free Physical Memory is 16.4Gb
2022-02-27T22:55:10.4356001Z 
2022-02-27T22:55:10.4356273Z Initialization took 42 seconds
2022-02-27T22:55:10.4359685Z Ready for connections!
2022-02-27T22:55:11.8159473Z Reading CustomSettings.config from Build
2022-02-27T22:55:14.8536964Z Skipping app 'C:\Applications\TestFramework\TestLibraries\permissions mock\Microsoft_Permissions Mock.app' as it is already installed
2022-02-27T22:55:14.8817596Z Skipping app 'C:\Applications\TestFramework\TestRunner\Microsoft_Test Runner.app' as it is already installed
2022-02-27T22:55:14.9071616Z Skipping app 'C:\Applications\TestFramework\TestLibraries\Any\Microsoft_Any.app' as it is already installed
2022-02-27T22:55:14.9330503Z Skipping app 'C:\Applications\TestFramework\TestLibraries\Assert\Microsoft_Library Assert.app' as it is already installed
2022-02-27T22:55:14.9604534Z Skipping app 'C:\Applications\TestFramework\TestLibraries\permissions mock\Microsoft_Permissions Mock.app' as it is already installed
2022-02-27T22:55:14.9929706Z Skipping app 'C:\Applications\TestFramework\TestLibraries\Variable Storage\Microsoft_Library Variable Storage.app' as it is already installed
2022-02-27T22:55:15.0216875Z Skipping app 'C:\Applications\system application\test\Microsoft_System Application Test Library.app' as it is already installed
2022-02-27T22:55:15.0592532Z Skipping app 'C:\Applications\BaseApp\Test\Microsoft_Tests-TestLibraries.app' as it is already installed
2022-02-27T22:55:15.0604115Z TestToolkit successfully imported
2022-02-27T22:55:15.4148602Z Container Build successfully created
2022-02-27T22:55:15.4165795Z 
2022-02-27T22:55:15.4210718Z Use:
2022-02-27T22:55:15.4289510Z Get-BcContainerEventLog -containerName Build to retrieve a snapshot of the event log from the container
2022-02-27T22:55:15.4338014Z Get-BcContainerDebugInfo -containerName Build to get debug information about the container
2022-02-27T22:55:15.4367169Z Enter-BcContainer -containerName Build to open a PowerShell prompt inside the container
2022-02-27T22:55:15.4379423Z Remove-BcContainer -containerName Build to remove the container again
2022-02-27T22:55:15.4391745Z docker logs Build to retrieve information about URL's again
2022-02-27T22:55:15.9227199Z ##[section]Finishing: Create build container

2022-02-27T22:55:15.9255585Z ##[section]Starting: Copy source into container folder
2022-02-27T22:55:15.9586693Z ==============================================================================
2022-02-27T22:55:15.9586943Z Task         : PowerShell
2022-02-27T22:55:15.9587115Z Description  : Run a PowerShell script
2022-02-27T22:55:15.9587269Z Version      : 1.2.3
2022-02-27T22:55:15.9587437Z Author       : Microsoft Corporation
2022-02-27T22:55:15.9587684Z Help         : [More Information](https://go.microsoft.com/fwlink/?LinkID=613736)
2022-02-27T22:55:15.9587968Z ==============================================================================
2022-02-27T22:55:15.9654035Z ##[command]. 'C:\Users\xxxxxxx\AppData\Local\Temp\e88df505-096a-4316-a244-6ab34d70adc7.ps1' 
2022-02-27T22:55:16.5363603Z 
2022-02-27T22:55:16.5366640Z 
2022-02-27T22:55:16.5371901Z     Directory: C:\ProgramData\BcContainerHelper\Extensions\Build
2022-02-27T22:55:16.5372896Z 
2022-02-27T22:55:16.5374055Z 
2022-02-27T22:55:16.5385492Z Mode                 LastWriteTime         Length Name                                                                 
2022-02-27T22:55:16.5389302Z ----                 -------------         ------ ----                                                                 
2022-02-27T22:55:16.5394043Z d-----         2/27/2022   5:55 PM                Source                                                               
2022-02-27T22:55:16.6877442Z 
2022-02-27T22:55:16.6877719Z 
2022-02-27T22:55:16.7045258Z ##[section]Finishing: Copy source into container folder`

Output: Compile-App:

`2022-02-27T22:55:16.7068146Z ##[section]Starting: Compile app
2022-02-27T22:55:16.7367776Z ==============================================================================
2022-02-27T22:55:16.7368023Z Task         : PowerShell
2022-02-27T22:55:16.7368194Z Description  : Run a PowerShell script
2022-02-27T22:55:16.7368349Z Version      : 1.2.3
2022-02-27T22:55:16.7368515Z Author       : Microsoft Corporation
2022-02-27T22:55:16.7368760Z Help         : [More Information](https://go.microsoft.com/fwlink/?LinkID=613736)
2022-02-27T22:55:16.7369031Z ==============================================================================
2022-02-27T22:55:16.7419323Z ##[command]. 'C:\Users\xxxxxxxx\AppData\Local\Temp\e85ebf39-63d0-4e93-851b-03a62912b46e.ps1' 
2022-02-27T22:55:17.5531587Z BcContainerHelper version 3.0.2
2022-02-27T22:55:17.5636300Z BcContainerHelper emits usage statistics telemetry to Microsoft
2022-02-27T22:55:20.5871132Z WARNING: The names of some imported commands from the module 'BcContainerHelper' include unapproved verbs that might 
2022-02-27T22:55:20.5872085Z make them less discoverable. To find the commands with unapproved verbs, run the Import-Module command again with the 
2022-02-27T22:55:20.5872454Z Verbose parameter. For a list of approved verbs, type Get-Verb.
2022-02-27T22:55:25.5675571Z Using Symbols Folder: C:\ProgramData\BcContainerHelper\Extensions\Build\Source\.alpackages
2022-02-27T22:55:41.9385830Z Downloading symbols: Microsoft_Application_19.0.29894.30693.app
2022-02-27T22:55:41.9424258Z Url : http://172.31.59.129:7049/BC/dev/packages?publisher=Microsoft&appName=Application&versionText=18.0.0.0&tenant=default
2022-02-27T22:55:44.2817874Z Adding dependency to System Application from Microsoft
2022-02-27T22:55:44.2846358Z Adding dependency to Base Application from Microsoft
2022-02-27T22:55:44.2887636Z Downloading symbols: Microsoft_System_19.0.29884.30666.app
2022-02-27T22:55:44.2896963Z Url : http://172.31.59.129:7049/BC/dev/packages?publisher=Microsoft&appName=System&versionText=18.0.0.0&tenant=default
2022-02-27T22:55:45.5256588Z Downloading symbols: xxxxxxxxxxxxxxxxxx_19.0.1.34.app <<-**Dependency from app.json**
2022-02-27T22:55:45.5276767Z Url : http://172.31.59.129:7049/BC/dev/packages/publisher=xxxxxxxxxxxxxx&appName=xxxxxxxxxxxxxxxxxxxxxxxxxxx&versionText=19.0.1.34&tenant=default <<--Still using Microsoft repository address
2022-02-27T22:55:45.9776394Z ERROR The remote server returned an error: (404) Not Found.
2022-02-27T22:55:47.5524422Z Compile-AppInNavContainer Telemetry Correlation Id: 60e58094-4514-448c-9e4c-4cb76e44e320
2022-02-27T22:55:47.6632607Z ##[error]The remote server returned an error: (404) Not Found. No published package matches the provided arguments.
At C:\Program Files\WindowsPowerShell\Modules\BcContainerHelper\3.0.2\AppHandling\Compile-AppInNavContainer.ps1:413 
char:21
+                     throw (GetExtendedErrorMessage $_)
+                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : OperationStopped: (The remote serv...ided arguments.:String) [], RuntimeException
    + FullyQualifiedErrorId : The remote server returned an error: (404) Not Found. No published package matches the p 
   rovided arguments.
 

2022-02-27T22:55:47.6863396Z ##[error]Process completed with exit code 1 and had 1 error(s) written to the error stream.
2022-02-27T22:55:47.6866950Z ##[section]Finishing: Compile app`

**Comment by freddydk** [comment]
Compile-AppInBcContainer has a -appSymbolsFolder, which you can provide and pre-populate with the apps you have a dependency on

**Comment by buzzwick** [comment]
Okay I will try that again. Does the appSymbolsFolder have to be shared with the container, or just with the script running Complie-AppInBCContainer? Thanks.

**Comment by freddydk** [comment]
It has to be shared with the container - you can share additional folders using -additionalParameters @("--volume ""c:\folderonhost:c:\folderincontainer""")
Then you specify the folder on the host as symbol folder.

**Comment by buzzwick** [comment]
Oh, thank you for that. “Sharing with the container” has a number of interpretations and methods in this space, and you have just truncated my pursuit of wrong ones. Now that I know that This Is How It Is Supposed to Work, I can persist until it does. Which I will do. Very much appreciate your indulgence.

**Comment by buzzwick** [comment]
Since you are closing this one, I will let you know how far I got: It took a lot of digging and searching, but I finally got all the variables and values in the right places. It was actually reading your documentation on Get-BcContainerPath that made the light go on. Now I can run the script, leave the container in place, query the same values as Compile-AppInNavContainer, 

`    $containerSymbolsFolder = Get-BcContainerPath -containerName $containerName -path $appSymbolsFolder
    if ("$containerSymbolsFolder" -eq "") {
        throw "The appSymbolsFolder ($appSymbolsFolder) is not shared with the container."
    }`

and confirm that they are correct. Shared folder and its contents is visible inside the container. Variables come back with correct values when queried interactively. But when Compile-AppInNavContainer runs, it still throws the error. So I commented out that query in Compile-AppInNavContainer, and got as far as getting another error further down the line:

`##[error]New-Item : Cannot find drive. A drive with the name '"C' does not exist.`

which suggests to me that maybe somewhere a path has been punctuated wrong.

Anyway, that is my next course--to try and isolate that and see if I can find what to fix in my script. I am new to this, so I did not want to keep posting questions until I really did run out of things to try on my own. Thanks.
